README
Gtranscode Readme
-----------------

Gtranscode is a program for "transcoding" video files.  This means it changes
the file format.  For example you can use it to change a .avi file to a .ogg
file.  This is useful as ogg will be more widely supported under linux systems.

This readme is in two parts, Installation and Use.

-------------------------------------------------------------------------------

I. Installation

 To build gtranscode extract the tarball:
   tar -zxf gtranscode*.tar.gz

 Change directory to the source:
   cd gtranscode*

 Configure the system:
   ./configure
 You can see configuration options by running
   ./configure --help

 Compile the program:
   make

 Install the program:
   make install

-------------------------------------------------------------------------------

II. Use

 To run gtranscode, first start the program
   gtranscode
 (if you have compiled the program but not installed it, you can run
   src/gtranscode from the source directory)

 Then you must select the file that you wish to transcode from.  The main dialog then appears.  First you must select the file format you want.  The most common formats are:

 AVI  - a widely used format under Windows and Linux.
 OGG  - a format designed for both flexibilty and freedom from patents.
 MPEG - This format is encumbered by patents but widely supported by software
        and hardware such as Digital TV recorders.

After selecting the target file format, the audio and video codecs must be selected.  These determine how the sound and video is represented digitally.

For Ogg, you almost certainly want Vorbis audio and Theora Video.

For Avi, you probably want mp3 audio if the option is available; AC-3 audio also works well.  MPEG-4 and XVID/DIVX good video codecs to use.

For MPEG you probably want mp3 audio and MPEG-2 Video.

When you have chosen all the settings, press start to begin the transcoding.
Gtranscode will ask you for a filename to save the new video as.  After choosing the process begins.  A progress bar is shown.  The process can be stopped at
any time with the stop button.


ChangeLog
gtranscode 0.1  2008-05-13

* Initial Release.  Features are:

* Any gstreamer-supported file input
* Most gstreamer-supported file formats and codecs
* Simple interface for transcoding


src/main.c
/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#define DEBUG TRUE

#include <gnome.h>
#include <glade/glade.h>
#include <gst/gst.h>
#include <gst/gstfilter.h>
#include <glib.h>
#include <glib/gstdio.h>
#include <libgnome/gnome-i18n.h>

#include "gtranscode.h"

// char* name, GstElementFactory* element_factory, GList* options ,GtkListStore allowed_audio_codecs, GtkListStore allowed_video_codecs)
GtkListStore * sources;
GtkListStore * containers;
gchar * source_filename;

GstElement *pipeline;

GladeXML *xml;

void
element_factory_add_to_gtk_list_store (GstElementFactory * element_factory, GtkListStore * group)
{
    GtkTreeIter iter;
    
    gtk_list_store_append( GTK_LIST_STORE (group), &iter);
    gtk_list_store_set ( GTK_LIST_STORE (group), &iter ,
                        0, gst_element_factory_get_longname (element_factory),
                        1, element_factory,
                        2, NULL,
                        3, NULL,
                        4, NULL, -1);
    /*initiase options*/
}

void
element_factory_add_to_gtk_list_store_with_children (GstElementFactory * element_factory, GtkListStore * group)
{
    GtkTreeIter iter;
    GtkListStore * audio_codecs_list_store;
    GtkListStore * video_codecs_list_store;
    GList * static_pad_templates = NULL;
    GList * pad_templates = NULL;
    GList * sink_pad_templates = NULL;
    GList * sink_caps = NULL;
    GList * audio_codecs = gst_registry_feature_filter( gst_registry_get_default(),
                                                       (GstPluginFeatureFilter) gtranscode_feature_filter_by_klass,
                                                       FALSE,
                                                       "Codec/Encoder/Audio");
    GList * video_codecs = gst_registry_feature_filter( gst_registry_get_default(),
                                                       (GstPluginFeatureFilter) gtranscode_feature_filter_by_klass,
                                                       FALSE,
                                                       "Codec/Encoder/Video");
    
    
    static_pad_templates = (GList *) gst_element_factory_get_static_pad_templates(element_factory);
    g_list_foreach (static_pad_templates,
                    (GFunc) gtranscode_static_pad_template_get_to_list,
                    &pad_templates);
    sink_pad_templates = gst_filter_run(pad_templates,
                                        (GstFilterFunc) gtranscode_pad_templates_is_sink,
                                        FALSE,
                                        NULL);
    g_list_foreach (sink_pad_templates,
                    (GFunc) gtranscode_pad_template_get_caps_to_list,
                    &sink_caps);
    
    
    audio_codecs = gst_filter_run (audio_codecs,
                                   (GstFilterFunc) gtranscode_element_factory_can_src_caps,
                                   FALSE,
                                   sink_caps);
    video_codecs = gst_filter_run (video_codecs,
                                   (GstFilterFunc) gtranscode_element_factory_can_src_caps,
                                   FALSE,
                                   sink_caps);
    
    if (g_list_length(audio_codecs) > 0 && g_list_length(video_codecs) > 0)
    {
        audio_codecs_list_store = gtk_list_store_new ( 5 ,  G_TYPE_STRING, G_TYPE_POINTER, G_TYPE_POINTER, G_TYPE_OBJECT, G_TYPE_OBJECT );
        video_codecs_list_store = gtk_list_store_new ( 5 ,  G_TYPE_STRING, G_TYPE_POINTER, G_TYPE_POINTER, G_TYPE_OBJECT, G_TYPE_OBJECT );
        g_list_foreach ( audio_codecs,
                        (GFunc) element_factory_add_to_gtk_list_store,
                        audio_codecs_list_store);
        g_list_foreach (video_codecs,
                        (GFunc) element_factory_add_to_gtk_list_store,
                        video_codecs_list_store);
        gtk_list_store_append( GTK_LIST_STORE (group), &iter);
        gtk_list_store_set ( GTK_LIST_STORE (group), &iter ,
                            0, gst_element_factory_get_longname (element_factory),
                            1, element_factory,
                            2, NULL,
                            3, audio_codecs_list_store,
                            4, video_codecs_list_store, -1);
    }
    
    g_list_free(static_pad_templates);
    g_list_free(pad_templates);
    g_list_free(sink_pad_templates);
    g_list_free(sink_caps);
    g_list_free(audio_codecs);
    g_list_free(video_codecs);
}

gboolean
gtranscode_pad_templates_is_sink (GstPadTemplate * pad_template, gpointer user_data)
{
    return (GST_PAD_TEMPLATE_DIRECTION(pad_template) == GST_PAD_SINK) ? TRUE : FALSE;
}

void
gtranscode_static_pad_template_get_to_list (GstStaticPadTemplate * static_pad_template,
                                            GList ** pad_template_list)
{
    *pad_template_list = g_list_append(*pad_template_list, gst_static_pad_template_get(static_pad_template));
}

void
gtranscode_pad_template_get_caps_to_list (GstPadTemplate * pad_template,
                                          GList ** caps_list)
{
    *caps_list = g_list_append(*caps_list, gst_pad_template_get_caps ( pad_template ));
}

gboolean
gtranscode_element_factory_can_src_caps (GstElementFactory * element_factory, GList * caps )
{
    do
    {
        if ( gst_element_factory_can_src_caps ( element_factory , caps->data ) == TRUE)
        {
            return TRUE;
        }
    }
    while ((caps = g_list_next(caps)) != NULL );
    return FALSE;
}

gboolean
gtranscode_feature_filter_by_klass (GstPluginFeature * feature, gchar * class)
{
    const gchar *klass;
    guint rank;
    
    if (!GST_IS_ELEMENT_FACTORY (feature))
        return FALSE;
    klass = gst_element_factory_get_klass (GST_ELEMENT_FACTORY (feature));
    if (g_strrstr (klass, class) == NULL )
    {
        return FALSE;
    }
    
    /* only select elements with autoplugging rank */
    rank = gst_plugin_feature_get_rank (feature);
    if (rank < GST_RANK_NONE)
        return FALSE;
    
    return TRUE;
}

int
main (int argc, char *argv[])
{
    /*GstRegistry *registry; */
    GtkWidget * file_chooser_dialog;
    GtkCellRenderer * text_renderer;
    GtkCellLayout * cell_layout;
#ifdef ENABLE_NLS
    bindtextdomain (GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
    textdomain (PACKAGE);
#endif
    gnome_init (PACKAGE, VERSION, argc, argv);
    glade_gnome_init ();
    gst_init (&argc, &argv);
    /*
     * The .glade filename should be on the next line.
     */
    xml = glade_xml_new (PACKAGE_SOURCE_DIR "/gtranscode.glade", NULL, NULL);
    /* This is important */
    glade_xml_signal_autoconnect (xml);
    
    sources = gtk_list_store_new ( 5 ,  G_TYPE_STRING, G_TYPE_POINTER, G_TYPE_POINTER , G_TYPE_OBJECT, G_TYPE_OBJECT);
    containers = gtk_list_store_new ( 5 ,  G_TYPE_STRING, G_TYPE_POINTER, G_TYPE_POINTER , G_TYPE_OBJECT, G_TYPE_OBJECT);
    
    /* Detect Gstreamer elements avilable */
    g_list_foreach (
                    gst_registry_feature_filter( gst_registry_get_default(),
                                                (GstPluginFeatureFilter) gtranscode_feature_filter_by_klass,
                                                FALSE,
                                                "Source/File"),
                    (GFunc) element_factory_add_to_gtk_list_store,
                    sources);
    g_list_foreach (
                    gst_registry_feature_filter( gst_registry_get_default(),
                                                (GstPluginFeatureFilter) gtranscode_feature_filter_by_klass,
                                                FALSE,
                                                "Codec/Muxer"),
                    (GFunc) element_factory_add_to_gtk_list_store_with_children ,
                    containers);
    
    /* Set up ui for Gstreamer elements available */
    text_renderer = gtk_cell_renderer_text_new();
    cell_layout = GTK_CELL_LAYOUT (glade_xml_get_widget (xml, "sources_combobox"));
    gtk_cell_layout_pack_start(cell_layout, text_renderer,TRUE);
    gtk_cell_layout_set_attributes ( cell_layout ,
                                    text_renderer,
                                    "text",
                                    0,NULL);
    
    text_renderer = gtk_cell_renderer_text_new();
    cell_layout = GTK_CELL_LAYOUT (glade_xml_get_widget (xml, "container_combobox"));
    gtk_cell_layout_pack_start(cell_layout, text_renderer,TRUE);
    gtk_cell_layout_set_attributes ( cell_layout ,
                                    text_renderer,
                                    "text",
                                    0,NULL);
    
    text_renderer = gtk_cell_renderer_text_new();
    cell_layout = GTK_CELL_LAYOUT (glade_xml_get_widget (xml, "audio_codec_combobox"));
    gtk_cell_layout_pack_start(cell_layout, text_renderer,TRUE);
    gtk_cell_layout_set_attributes ( cell_layout ,
                                    text_renderer,
                                    "text",
                                    0,NULL);
    
    text_renderer = gtk_cell_renderer_text_new();
    cell_layout = GTK_CELL_LAYOUT (glade_xml_get_widget (xml, "video_codec_combobox"));
    gtk_cell_layout_pack_start(cell_layout, text_renderer,TRUE);
    gtk_cell_layout_set_attributes ( cell_layout ,
                                    text_renderer,
                                    "text",
                                    0,NULL);
    
    gtk_combo_box_set_model ( GTK_COMBO_BOX (glade_xml_get_widget (xml, "sources_combobox")),
                             GTK_TREE_MODEL (sources));
    gtk_combo_box_set_model ( GTK_COMBO_BOX (glade_xml_get_widget (xml, "container_combobox")),
                             GTK_TREE_MODEL (containers));
    
    file_chooser_dialog = gtk_file_chooser_dialog_new ("Open media file",
                                                       NULL,
                                                       GTK_FILE_CHOOSER_ACTION_OPEN,
                                                       GTK_STOCK_OPEN, GTK_RESPONSE_ACCEPT,
                                                       NULL);
  if (gtk_dialog_run (GTK_DIALOG (file_chooser_dialog)) == GTK_RESPONSE_ACCEPT)
  {
    source_filename = gtk_file_chooser_get_filename (GTK_FILE_CHOOSER (file_chooser_dialog));
    gtk_widget_hide (file_chooser_dialog);
    gtk_widget_show (glade_xml_get_widget (xml, "gtranscode_app"));
    gtk_main ();
  }
  else
  {
      exit (1);
  }
    return 0;
}


src/gtranscode.h
/***************************************************************************
 *            gtranscode.h
 *
 *  Sun Jan 27 22:59:09 2008
 *  Copyright  2008  David Newgas
 *  david@newgas.net
 ****************************************************************************/

/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifndef _GTRANSCODE_H
#define _GTRANSCODE_H

void gtranscode_object_set (gchar ** option_pair, gpointer object);

void
element_factory_add_to_gtk_list_store (GstElementFactory * element_factory,
                                       GtkListStore * group);

void
element_factory_add_to_gtk_list_store_with_children (GstElementFactory * element_factory,
                                                     GtkListStore * group);

GList *
gtranscode_element_factory_get_options (GstElementFactory * element_factory);

void
gtranscode_static_pad_template_get_to_list (GstStaticPadTemplate * static_pad_template,
                                            GList ** pad_template_list);

gboolean
gtranscode_pad_templates_is_sink (GstPadTemplate * pad_template,
                                  gpointer user_data);

void
gtranscode_static_pad_template_get_to_list (GstStaticPadTemplate * static_pad_template,
                                            GList ** pad_template_list);

void
gtranscode_pad_template_get_caps_to_list (GstPadTemplate * pad_template,
                                          GList ** caps_list);

gboolean
gtranscode_element_factory_can_src_caps (GstElementFactory * element_factory,
                                         GList * caps );

void
pad_added (GstElement * element1,
           GstPad * pad, gboolean arg1,
           GstElement * element2);

GstElement * transcode (GstElementFactory * source_factory,
                        GList * source_opts,
                        GstElementFactory * container_factory,
                        GList * container_opts,
                        GstElementFactory * audio_codec_factory,
                        GList * audio_codec_opts,
                        GstElementFactory * video_codec_factory,
                        GList * video_codec_opts,
                        GstElementFactory * sink_factory,
                        GList * sink_opts);

void gtranscode_ui_update (GtkComboBox * combo_box);

void transcode_button_clicked (GtkButton * button);

void
gtranscode_options_button_clicked (GtkButton * button, GtkComboBox * widget);

int main (int argc, char *argv[]);

void
gtranscode_message_eos (GstBus * bus, GstMessage * message, gpointer data);

void
gtranscode_message_error (GstBus * bus, GstMessage * message, gpointer data);

void
gtranscode_stop_button_clicked (GtkButton * button);

void
gtranscode_message_state_changed (GstBus * bus, GstMessage *, gpointer data);

gboolean
gtranscode_feature_filter_by_klass (GstPluginFeature * feature, gchar * class);
#endif /* _GTRANSCODE_H */


src/gtranscode-ui-handlers.c
/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <gnome.h>
#include <glade/glade.h>
#include <glib.h>
#include <glib/gstdio.h>
#include <gst/gst.h>

#include "gtranscode.h"

GstElement * pipeline;
GtkListStore * sources, *containers;
GladeXML * xml;
gchar * source_filename;

void
gtranscode_ui_update (GtkComboBox * combo_box)
{
    GtkTreeIter iter;
    GtkListStore* list_store;
    GValue value = {0, };
    gtk_combo_box_get_active_iter ( combo_box , &iter );
    gtk_tree_model_get_value ( GTK_TREE_MODEL (containers), &iter, 3, &value);
    list_store = g_value_get_object(&value);
    g_value_unset(&value);
    gtk_combo_box_set_model ( GTK_COMBO_BOX (glade_xml_get_widget (xml, "audio_codec_combobox")),
                             GTK_TREE_MODEL (list_store));
    gtk_tree_model_get_value ( GTK_TREE_MODEL (containers), &iter, 4, &value);
    list_store = g_value_get_object(&value);
    g_value_unset(&value);
    gtk_combo_box_set_model ( GTK_COMBO_BOX (glade_xml_get_widget (xml, "video_codec_combobox")),
                             GTK_TREE_MODEL (list_store));
}


gboolean gtranscode_ui_update_position (gpointer data)
{
    gint64 pos, len;
    GstFormat fmt = GST_FORMAT_TIME;
    if (!GST_IS_ELEMENT(pipeline))
    {
        return FALSE;
    }
    if (gst_element_query_position (pipeline, &fmt, &pos)
        && gst_element_query_duration (pipeline, &fmt, &len)) {
            gnome_appbar_set_progress_percentage ( GNOME_APPBAR(glade_xml_get_widget(xml,"appbar")),
                                                  (gfloat) pos / (gfloat) len);
        }
    
    /* call me again */
    return TRUE;
}
void
gtranscode_stop_button_clicked (GtkButton * button)
{
    gst_element_set_state (pipeline, GST_STATE_NULL);
    gst_object_unref(pipeline);
    gnome_appbar_set_progress_percentage ( GNOME_APPBAR(glade_xml_get_widget(xml,"appbar")),
                                                  0.0);
    gtk_widget_hide ( glade_xml_get_widget (xml, "stop-button"));
    gtk_widget_show ( glade_xml_get_widget (xml, "start-button"));
    /* The gtk functions should really be called in gtranscode_message_state_changed
     but that doesn't seem to work...!!*/
}

void
gtranscode_transcode_button_clicked (GtkButton * button)
{
  GstElementFactory *source_factory,
  *container_factory,
  *audio_codec_factory, *video_codec_factory, *sink_factory;
  GstElement * pipeline;
  GstBus * bus;
  GValue value = {0, };
  GtkTreeIter iter;
  GList *source_opts,
        *container_opts,
        *audio_codec_opts,
        *video_codec_opts,
        *sink_opts;
  gchar *filesinkopts[] = {"location", NULL};
  gchar *filesrcopts[] = {
    "location",
    source_filename
  };
  GtkWidget * file_chooser_dialog = gtk_file_chooser_dialog_new ("Save as",
                                                                 NULL,
                                                                 GTK_FILE_CHOOSER_ACTION_SAVE,
                                                                 GTK_STOCK_OPEN, GTK_RESPONSE_ACCEPT,
                                                                 NULL);
  if (gtk_dialog_run (GTK_DIALOG (file_chooser_dialog)) == GTK_RESPONSE_ACCEPT)
  {
    filesinkopts[1] = gtk_file_chooser_get_filename (GTK_FILE_CHOOSER (file_chooser_dialog));
    gtk_widget_hide (file_chooser_dialog);
  }
  else
  {
    return;
  }
  gtk_combo_box_get_active_iter (GTK_COMBO_BOX( glade_xml_get_widget (xml, "sources_combobox")),
                                 &iter );
  gtk_tree_model_get_value ( GTK_TREE_MODEL (sources), &iter, 1, &value);
  source_factory = g_value_get_pointer( &value);
  source_factory = gst_element_factory_find ("filesrc");
  source_opts = NULL;
  source_opts = g_list_append (source_opts, filesrcopts);


  gtk_combo_box_get_active_iter (GTK_COMBO_BOX( glade_xml_get_widget (xml, "container_combobox")),
                                 &iter );
  g_value_unset(&value);
  gtk_tree_model_get_value ( GTK_TREE_MODEL (containers), &iter, 1, &value);
  container_factory = g_value_get_pointer( &value);
  container_opts = NULL;


  gtk_combo_box_get_active_iter (GTK_COMBO_BOX( glade_xml_get_widget (xml, "audio_codec_combobox")),
                                 &iter );
  g_value_unset(&value);
  gtk_tree_model_get_value (gtk_combo_box_get_model(GTK_COMBO_BOX( glade_xml_get_widget (xml, "audio_codec_combobox"))),
                            &iter, 1, &value);
  audio_codec_factory = g_value_get_pointer( &value);
  audio_codec_opts = NULL;


  gtk_combo_box_get_active_iter (GTK_COMBO_BOX( glade_xml_get_widget (xml, "video_codec_combobox")),
                                 &iter );
  g_value_unset(&value);
  gtk_tree_model_get_value ( gtk_combo_box_get_model(GTK_COMBO_BOX( glade_xml_get_widget (xml, "video_codec_combobox"))),
                            &iter, 1, &value);
  video_codec_factory = g_value_get_pointer( &value);
  video_codec_opts = NULL;


  sink_factory = gst_element_factory_find ("filesink");
  sink_opts = NULL;
  sink_opts = g_list_append (sink_opts, filesinkopts);
  pipeline = transcode (source_factory, source_opts, container_factory, container_opts,
                        audio_codec_factory, audio_codec_opts, video_codec_factory,
                        video_codec_opts, sink_factory, sink_opts);

  bus = gst_pipeline_get_bus (GST_PIPELINE (pipeline));
  gst_bus_add_signal_watch (bus);
  g_signal_connect (bus, "message::error", G_CALLBACK (gtranscode_message_error), NULL);
  g_signal_connect (bus, "message::eos", G_CALLBACK (gtranscode_message_eos), NULL);
  g_signal_connect (bus, "message::state-changed", G_CALLBACK( gtranscode_message_state_changed),NULL);
  gst_object_unref(bus);

  g_timeout_add (200, (GSourceFunc) gtranscode_ui_update_position, NULL);

  gst_element_set_state (pipeline, GST_STATE_PLAYING);
}

void
gtranscode_options_button_clicked (GtkButton * button, GtkComboBox * widget)
{
    GtkTreeIter iter;
    GstElementFactory * element_factory;
    GValue value = {0, };
    GList *options = NULL;
    if (gtk_combo_box_get_active_iter (widget,
                                       &iter ) == FALSE )
    {
        return;
    }
    gtk_tree_model_get_value (gtk_combo_box_get_model(widget),
                              &iter, 2, &value);
    options = g_value_get_pointer(&value);
    g_value_unset(&value);
    gtk_tree_model_get_value (gtk_combo_box_get_model(widget),
                              &iter, 1, &value);
    element_factory = g_value_get_pointer(&value);
    
    /*initiase options in ui*/
    gtk_label_set_text (GTK_LABEL (glade_xml_get_widget(xml, "optionslabel")),
                        g_strconcat("Set options for ",
                                    gst_element_factory_get_longname( element_factory),
                                    NULL));
    gtk_widget_show (glade_xml_get_widget (xml, "options_dialog"));
}


src/gtranscode-gst-handlers.c
/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <gst/gst.h>
#include <gnome.h>
#include <glade/glade.h>

#include "gtranscode.h"

GstElement * pipeline;
GtkListStore * sources, *containers;
GladeXML * xml;

void
gtranscode_message_eos (GstBus * bus, GstMessage * message, gpointer data)
{
    GtkMessageDialog * dialog;
    gst_element_set_state (pipeline, GST_STATE_NULL);
    gst_object_unref (pipeline);
    dialog = GTK_MESSAGE_DIALOG (gtk_message_dialog_new (GTK_WINDOW( glade_xml_get_widget(xml,"gtranscode_app")),
                                                         GTK_DIALOG_DESTROY_WITH_PARENT,
                                                         GTK_MESSAGE_INFO,
                                                         GTK_BUTTONS_CLOSE,
                                                         "Transcoding Complete!"));
    gtk_dialog_run (GTK_DIALOG (dialog));
    gtk_widget_destroy (GTK_WIDGET(dialog));
}

void
gtranscode_message_error (GstBus * bus, GstMessage * message, gpointer data)
{
    GError *err;
    gchar *debug;
    GtkMessageDialog * dialog;
    gst_element_set_state (pipeline, GST_STATE_NULL);
    gst_object_unref (pipeline);
    gst_message_parse_error (message, &err, &debug);
    dialog = GTK_MESSAGE_DIALOG(gtk_message_dialog_new (GTK_WINDOW(glade_xml_get_widget(xml,"gtranscode_app")),
                                                        GTK_DIALOG_DESTROY_WITH_PARENT,
                                                        GTK_MESSAGE_ERROR,
                                                        GTK_BUTTONS_CLOSE,
                                                        "Error transcoding: %s",
                                                        err->message));
    gtk_dialog_run (GTK_DIALOG (dialog));
    gtk_widget_destroy (GTK_WIDGET(dialog));
}

void
gtranscode_message_state_changed (GstBus * bus, GstMessage * message, gpointer data)
{
    GstState oldstate, newstate, pendingstate;
    gst_message_parse_state_changed ( message, &oldstate, &newstate, &pendingstate);
    switch( newstate )
    {
        case GST_STATE_READY:
        case GST_STATE_PAUSED:	
            gst_element_set_state( pipeline, GST_STATE_PLAYING);		  
            break;
        case GST_STATE_PLAYING:
            gtk_widget_hide ( glade_xml_get_widget (xml, "start-button"));
            gtk_widget_show ( glade_xml_get_widget (xml, "stop-button"));
            break;
        case GST_STATE_VOID_PENDING:
            break;
        case GST_STATE_NULL:
            gtk_widget_hide ( glade_xml_get_widget (xml, "stop-button"));
            gtk_widget_show ( glade_xml_get_widget (xml, "start-button"));
    }
}


src/gtranscode-transcode.c
/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#define DEBUG TRUE

#include <gst/gst.h>
#include <glib.h>
#include <gnome.h>
#ifdef DEBUG
#include <glib/gstdio.h>
#endif

#include "gtranscode.h"

GstElement *pipeline;

void
gtranscode_object_set (gchar ** option_pair, gpointer object)
{
    g_object_set (G_OBJECT (object), option_pair[0], option_pair[1], NULL);
}

void
pad_added (GstElement
           * src, GstPad * src_pad, gboolean arg1, GstElement * sink)
{
    GstPad *sink_pad;
    GstPadLinkReturn success = GST_PAD_LINK_OK;
    sink_pad = gst_element_get_pad (sink, "sink");
    
    if (gst_pad_can_link (src_pad, sink_pad))
    {
        success = gst_pad_link (src_pad, sink_pad);
    }
    return;
}

GstElement *
transcode
(GstElementFactory *
 source_factory,
 GList * source_opts,
 GstElementFactory *
 container_factory,
 GList *
 container_opts,
 GstElementFactory *
 audio_codec_factory,
 GList *
 audio_codec_opts,
 GstElementFactory *
 video_codec_factory,
 GList *
 video_codec_opts, GstElementFactory * sink_factory, GList * sink_opts)
{
    GstElement *source, *container, *video_codec, *audio_convert, *audio_codec, *sink, *decoder;
    /* Set up gst elements */
    pipeline = gst_pipeline_new ("transcoder");
    source = gst_element_factory_create (source_factory, "source");
    decoder = gst_element_factory_make ("decodebin2", "decoder");
    container = gst_element_factory_create (container_factory, "muxer");
    video_codec =
        gst_element_factory_create (video_codec_factory, "video_codec");
    audio_convert = gst_element_factory_make ("audioconvert" , "audio_convert");
    audio_codec =
        gst_element_factory_create (audio_codec_factory, "audio_codec");
    sink = gst_element_factory_create (sink_factory, "sink");
    if (!pipeline || !audio_convert
        || !source || !container || !video_codec || !audio_codec || !sink)
    {
        g_printerr ("Could not initialise elements!\n");
    }
    
    g_list_foreach (source_opts, (GFunc) gtranscode_object_set, source);
    g_list_foreach (container_opts, (GFunc) gtranscode_object_set, container);
    g_list_foreach
        (video_codec_opts, (GFunc) gtranscode_object_set, video_codec);
    g_list_foreach
        (audio_codec_opts, (GFunc) gtranscode_object_set, audio_codec);
    g_list_foreach (sink_opts, (GFunc) gtranscode_object_set, sink);
    gst_bin_add_many
        (GST_BIN (pipeline),audio_convert,
         source, decoder, video_codec, audio_codec, container, sink, NULL);
    gst_element_link_many (video_codec, container, sink, NULL);
    gst_element_link_many (audio_convert, audio_codec, container, NULL);
    gst_element_link_many (source, decoder, NULL);
    g_signal_connect
        (decoder, "new-decoded-pad", G_CALLBACK (pad_added), video_codec);
    g_signal_connect
        (decoder, "new-decoded-pad", G_CALLBACK (pad_added), audio_convert);
    /* Set up events for when it's done */
    /* Set up progress indicator */
    return pipeline;
}
